contrib/win32/win32compat/misc.c

```diff
@@ -1587,55 +1587,122 @@ am_system()
+ * Sets psid_len (if non-NULL) to the actual SID size.
+ * Caller should free() return value if psid argument was NULL.
  */
 PSID
-get_sid(const char* name)
+lookup_sid(const wchar_t* name_utf16, PSID psid, DWORD * psid_len)
 {
-	HANDLE token = NULL;
-	TOKEN_USER* info = NULL;
-	DWORD info_len = 0;
-	PSID ret = NULL, psid = NULL;
-	wchar_t* name_utf16 = NULL;
+	PSID ret = NULL, alloc_psid = NULL, target_psid;
+	DWORD sid_len = 0;
+	SID_NAME_USE n_use;
+	WCHAR dom[DNLEN + 1] = L"";
+	DWORD dom_len = DNLEN + 1;
+	wchar_t* name_utf16_modified = NULL;
+	BOOL resolveAsAdminsSid = 0, r;
 
-	if (name) {
-		DWORD sid_len = 0;
-		SID_NAME_USE n_use;
-		WCHAR dom[DNLEN + 1] = L"";
-		DWORD dom_len = DNLEN + 1;
-		BOOL resolveAsAdminsSid = 0, r;
+	debug3_f("name_utf16:%S", name_utf16);
 
-		if ((name_utf16 = utf8_to_utf16(name)) == NULL)
-			goto cleanup;
+	LookupAccountNameW(NULL, name_utf16, NULL, &sid_len, dom, &dom_len, &n_use);
+
+	if (sid_len == 0 && _wcsicmp(name_utf16, L"administrators") == 0) {
+		CreateWellKnownSid(WinBuiltinAdministratorsSid, NULL, NULL, &sid_len);
+		resolveAsAdminsSid = 1;
+		debug3_f("resolveAsAdminsSid:%d", resolveAsAdminsSid);
+	}
 
-		LookupAccountNameW(NULL, name_utf16, NULL, &sid_len, dom, &dom_len, &n_use);
+	if (sid_len == 0) {
+		error_f("LookupAccountNameW() failed with error:%d", GetLastError());
+		errno = errno_from_Win32LastError();
+		goto cleanup;
+	}
 
-		if (sid_len == 0 && _stricmp(name, "administrators") == 0) {
-			CreateWellKnownSid(WinBuiltinAdministratorsSid, NULL, NULL, &sid_len);
-			resolveAsAdminsSid = 1;
+	target_psid = psid;
+	if (target_psid == NULL) {
+		if ((alloc_psid = malloc(sid_len)) == NULL) {
+			errno = ENOMEM;
+			error_f("Failed to allocate memory");
+			goto cleanup;
 		}
+		target_psid = alloc_psid;
+	}
 
-		if (sid_len == 0) {
-			errno = errno_from_Win32LastError();
+	if (resolveAsAdminsSid)
+		r = CreateWellKnownSid(WinBuiltinAdministratorsSid, NULL, target_psid, &sid_len);
+	else
+		r = LookupAccountNameW(NULL, name_utf16, target_psid, &sid_len, dom, &dom_len, &n_use);
+
+	if (!r) {
+		error_f("Failed to retrieve SID for user:%S error:%d", name_utf16, GetLastError());
+		errno = errno_from_Win32LastError();
+		goto cleanup;
+	}
+
+	if (n_use == SidTypeDomain) {
+		// Additionally check the case when name is the same as computer name and
+		// thus same as local domain. Try to resolve <name>\<name>.
+		/* fetch the computer name so we can determine if the specified user is local or not */
+		wchar_t computer_name[CNLEN + 1];
+		DWORD computer_name_size = ARRAYSIZE(computer_name);
+		if (GetComputerNameW(computer_name, &computer_name_size) == 0) {
+			error_f("GetComputerNameW() failed with error:%d", GetLastError());
 			goto cleanup;
 		}
 
-		if ((psid = malloc(sid_len)) == NULL) {
-			errno = ENOMEM;
+		if (_wcsicmp(name_utf16, computer_name) != 0) {
+			error_f("For SidTypeDomain, name:%ls must be same as machine name:%ls", name_utf16, computer_name);
```

Comment:
The wording here is little bit misleading for the case when the user actually tried to login with a domain name instead of full user name (by mistake/typo).

Reply:
I didn't get your comment.

If it's a domain user then we don't come into this block? n_use != SidTypeDomain for domain users?
We do support all forms of domain user format (ssh domain\user@localhost, ssh user@domain@localhost, etc).

Reply:
I meant if the user uses _only_ domain name to log in:

```
domain
```

instead of

```
domain\user
```

In this case I think (not tried) that the error “For SidTypeDomain, name:domain must be same as machine name:xxx” will be displayed. And this will be confusing.